opm/simulators/linalg/ISTLSolverEbos.hpp

```diff
@@ -678,87 +331,34 @@ DenseMatrix transposeDenseMatrix(const DenseMatrix& M)
-                if (this->iterations() > 10) {
-                    recreate_solver = true;
-                }
-            } else {
-                assert(this->parameters_.cpr_reuse_setup_ == 3);
-                assert(recreate_solver == false);
-                // Never recreate solver.
-            }
-
-            std::function<Vector()> weightsCalculator;
-
-            auto preconditionerType = prm_.get("preconditioner.type", "cpr");
-            if( preconditionerType  == "cpr" ||
-                preconditionerType == "cprt"
-                )
-            {
-                bool transpose = false;
-                if(preconditionerType == "cprt"){
-                    transpose = true;
-                }
 
-                auto weightsType = prm_.get("preconditioner.weight_type", "quasiimpes");
-                auto pressureIndex = this->prm_.get("preconditioner.pressure_var_index", 1);
-                if(weightsType == "quasiimpes") {
-                    // weighs will be created as default in the solver
-                    weightsCalculator =
-                        [this, transpose, pressureIndex](){
-                            return Opm::Amg::getQuasiImpesWeights<Matrix,
-                                                                  Vector>(getMatrix(),
-                                                                          pressureIndex,
-                                                                          transpose);
-                        };
-
-                }else if(weightsType == "trueimpes"  ){
-                    weightsCalculator =
-                        [this](){
-                            return this->getStorageWeights();
-                        };
-                }else{
-                    OPM_THROW(std::invalid_argument, "Weights type " << weightsType << "not implemented for cpr."
-                              << " Please use quasiimpes or trueimpes.");
-                }
-            }
+            std::function<Vector()> weightsCalculator = getWeightsCalculator();
 
-            if (recreate_solver || !flexibleSolver_) {
+            if (shouldCreateSolver()) {
                 if (isParallel()) {
 #if HAVE_MPI
                     if (useWellConn_) {
-                        assert(noGhostMat_);
                         using ParOperatorType = Dune::OverlappingSchwarzOperator<Matrix, Vector, Vector, Comm>;
                         linearOperatorForFlexibleSolver_ = std::make_unique<ParOperatorType>(getMatrix(), *comm_);
                         flexibleSolver_ = std::make_unique<FlexibleSolverType>(*linearOperatorForFlexibleSolver_, *comm_, prm_, weightsCalculator);
                     } else {
-                        if (!ownersFirst_) {
-                            OPM_THROW(std::runtime_error, "In parallel, the flexible solver requires "
-                                      "--owner-cells-first=true when --matrix-add-well-contributions=false is used.");
```

Comment:
Why isn't this needed anymore? ownersFirst is now alway false with your changes.

Reply:
See above comment, I meant for it to be true always.